### PR TITLE
Selection DOM interface (but not actual UI selections)

### PR DESCRIPTION
**Pull request by pshaughn**
This is work towards #7492.

I put new tests in the mozilla-specific directory rather than the main tree because I'm not sure how upstreamable they are; I'd like opinions on that point.

This adds a new exposed interface, which I understand from tests/mozilla/interfaces.html is something requiring specific approval from someone allowed to give that approval.

Things that aren't done here:
- The spec doesn't describe what selection/script-and-style-elements.html wants, and what it wants seems to require some sort of layout query.
- Actual UI interactions are not present at all; selection.rs has a few TODOs saying which methods I believe the eventual UI code should call for what actions, but there are a lot of fine points here that I don't know about.
- I haven't touched Node; you can ask if a node's in the Selection's Range the same way you'd ask about any other Range, but there's not a faster path just for selection layout.

### CODE REVIEW

File: components/script/dom/selection.rs [hunk 1]
```diff
@@ -0,0 +1,523 @@
...
+            return;
+        }
+        let this = Trusted::new(self);
+        let window = window_from_node(&*self.document);
+        window
+            .task_manager()
+            .dom_manipulation_task_source()
+            .queue(
+                task!(selectionchange_task_steps: move || {
+                let this = this.root();
```
Comment (pshaughn) [inline components/script/dom/selection.rs]: this is indented wrong, weird that tidy/fmt didn't notice

File: components/script/dom/selection.rs [hunk 2]
```diff
@@ -0,0 +1,523 @@
...
+
+    // https://w3c.github.io/selection-api/#dom-selection-selectallchildren
+    fn SelectAllChildren(&self, node: &Node) -> ErrorResult {
+        if node.is_doctype() {
+            // Range::new assumes its caller already checked this,
+            // WPT implies the check comes before the same-root check
+            return Err(Error::InvalidNodeType);
+        }
+        if !self.is_same_root(node) {
+            // Spec doesn't say to empty here, but WPT tests expect it.
```
Comment (pshaughn) [inline components/script/dom/selection.rs]: This comment is in error, just noting that here to remind myself

File: components/script/dom/selection.rs [hunk 3]
```diff
@@ -0,0 +1,523 @@
...
+            return false;
+        }
+    }
+
+    // https://w3c.github.io/selection-api/#dom-selection-stringifier
+    fn Stringifier(&self) -> DOMString {
+        // The spec as of Jan 31 2020 just says
+        // "See W3C bug 10583." for this method.
+        // Stringifying the range seems at least approximately right
+        // and passes the one case in the tentative WPT test.
```
Comment (pshaughn) [inline components/script/dom/selection.rs]: There's two cases; we fail the other one because it cares about CSS state

File: components/script/dom/selection.rs [hunk 4]
```diff
@@ -0,0 +1,523 @@
...
+        if anchor_offset > anchor_node.len() || focus_offset > focus_node.len() {
+            return Err(Error::IndexSize);
+        }
+
+        // Step 2
+        if !self.is_same_root(anchor_node) || !self.is_same_root(focus_node) {
+            return Ok(());
+        }
+
+        if anchor_node.is_doctype() || focus_node.is_doctype() {
```
Comment (pshaughn) [inline components/script/dom/selection.rs]: Looking at the other methods that are doing this, this should probably be before the IndexSize check, not after the same_root one.

**Comment by pshaughn** [comment]
Nothing too unexpected here... 

▶ Unexpected subtest result in /custom-elements/reactions/Selection.html:
  └ PASS [expected FAIL] deleteFromDocument on Selection must enqueue a disconnected reaction
  ▶ Unexpected subtest result in /html/browsers/the-window-object/window-properties.https.html:
  └ PASS [expected FAIL] Window method: getSelection
  ▶ Unexpected subtest result in /dom/ranges/Range-mutations-deleteData.html:
[many cases, probably the entire file like the other Range-mutations-* ones]
▶ Unexpected subtest result in /_mozilla/mozilla/interfaces.html:
  │ FAIL [expected PASS] Interfaces exposed on the window
  │   → assert_true: If this is failing: DANGER, are you sure you want to expose the new interface Selection to all webpages as a property on the global? Do not make a change to this file without review from jdm or Ms2ger for that specific change! expected true got false

**Comment by pshaughn** [comment]
```
1 unexpected results that are NOT known-intermittents:
  ▶ TIMEOUT [expected OK] /2dcontext/shadows/shadowBlur_gaussian_tolerance.1.html
  │ 
  │ 
  │ 
  └ 
```
I've never seen this one before. It looks like the test is iterating over many of image pixels one at a time in Javascript, so maybe it's susceptible to timeouts if it's not getting as many CPU cycles. It isn't touching selections or ranges.

**Comment by CYBAI** [comment]
@pshaughn I guess that's #25523 👀

File: components/script/dom/selection.rs [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+                return Err(Error::InvalidNodeType);
+            }
+
+            if offset > node.len() {
+                // As with is_doctype, not explicit in selection spec steps
+                // implied by range-setting
+                return Err(Error::IndexSize);
+            }
+
+            // let is_old_anchor_before_or_equal;
```
Comment (jdm) [inline components/script/dom/selection.rs]: Why is this here?

File: components/script/dom/selection.rs [hunk 6]
```diff
@@ -0,0 +1,522 @@
...
+    // https://w3c.github.io/selection-api/#dom-selection-setbaseandextent
+    fn SetBaseAndExtent(
+        &self,
+        anchor_node: &Node,
+        anchor_offset: u32,
+        focus_node: &Node,
+        focus_offset: u32,
+    ) -> ErrorResult {
+        // Step 1
+        if anchor_node.is_doctype() || focus_node.is_doctype() {
```
Comment (jdm) [inline components/script/dom/selection.rs]: Presumably this check comes from https://dom.spec.whatwg.org/#concept-range-bp-set? I'm a bit concerned that we're adding these extra checks in various places in the code; is our implementation missing them elsewhere, or is this a spec inconsistency, or are we working around tests that assume a certain order or checks that doesn't match what the spec says?

**Review by jdm — CHANGES_REQUESTED** [review]
Great work!

File: components/script/dom/selection.rs (hunk 6 above)
Comment (pshaughn) [reply components/script/dom/selection.rs]: Tests, which Firefox and Chrome both pass in FYI, are assuming a certain order that doesn't match what the spec says.

File: components/script/dom/selection.rs (hunk 5 above)
Comment (pshaughn) [reply components/script/dom/selection.rs]: Just forgetfulness.

File: components/script/dom/selection.rs (hunk 6 above)
Comment (pshaughn) [reply components/script/dom/selection.rs]: I brought it up here: https://github.com/w3c/selection-api/issues/118

**Comment by jdm** [comment]
@bors-servo retry
* #23290

**Comment by jdm** [comment]
@bors-servo retry
* #23290
